Add hint to `NotInScopeW` for visually confusable Unicode characters

mkhakimova42 opened the pull request:
Introduces a new hint to the `NotInScopeW` error during scope checking if the name that is out of scope is different from a name in scope only by visually similar characters (https://github.com/agda/agda/issues/5629 and https://github.com/agda/agda/issues/2898). Uses the [text-icu](https://hackage.haskell.org/package/text-icu-0.8.0.5) package to determine if characters are visually similar.

This PR was a part of my [Master's thesis](https://repository.tudelft.nl/record/uuid:52513287-7149-41f1-a8e8-8e38696cb283).

## Example
```agda
id : Set → Set
id bαp = baρ
```
```
Not in scope:
  b𝕒ρ at example.agda:2.10-13
    (did you accidentally use a confusable character?
       You typed:    b𝕒ρ
       In scope:     bαp
       (diff)        -^^
       
              Character        Character name                        agda-mode input  Emacs input      
       ---->  ‎'𝕒‎' (0x1d552)    MATHEMATICAL DOUBLE-STRUCK SMALL A    \ba              C-x 8 RET 1d552
          vs  ‎'α‎' (0x3b1)      GREEK SMALL LETTER ALPHA              \Ga              C-x 8 RET 3b1  
                                                                                                       
       ---->  ‎'ρ‎' (0x3c1)      GREEK SMALL LETTER RHO                \Gr              C-x 8 RET 3c1  
          vs  ‎'p‎' (0x70)       LATIN SMALL LETTER P                                   C-x 8 RET 70   
                                                                                                       )
when scope checking b𝕒ρ
```

## Future work
- To tell the user how to type a character, the code currently uses the `keybindings.json` file introduced in https://github.com/agda/agda/pull/7094. This file now contains inconsistencies with the actual keybindings. In future work, the keybinding info should probably be pulled directly from the `.el` file.

- By default, text-icu uses ICU's bundled copy of [confusables.txt](http://unicode.org/Public/security/latest/confusables.txt) which excludes some characters we might want to be considered visually similar (e.g. `::` vs `∷`). In the future, we could supply our own `confusables.txt` file.

mkhakimova42 committed:
Basic warning with suggested hexcode

mkhakimova42 committed:
Change table format and fix direction when RTL unicode used

mkhakimova42 committed:
Use Text.PrettyPrint.Boxes to print hint as table

mkhakimova42 committed:
Attempt at making table more readable

mkhakimova42 committed:
Preparations for loading keybindings JSON

mkhakimova42 committed:
Use JSON with keybindings to display agda-mode input for unicode

mkhakimova42 committed:
Restructure hint presentation for easier comparison of characters

mkhakimova42 committed:
Bring branch up to date with the uesr study version

noughtmare commented on src/full/Agda/TypeChecking/Pretty/Warning.hs:
```diff
@@ -685,6 +701,145 @@ didYouMean inscope canon x
+  getConfusChars xs ('_':ys) charsX charsY indicatorX indicatorY = getConfusChars xs ys charsX charsY indicatorX (indicatorY ++ "-")
+  getConfusChars (x:xs) (y:ys) charsX charsY indicatorX indicatorY
+    | x /= y = getConfusChars xs ys (charsX ++ [x]) (charsY ++ [y]) (indicatorX ++ "^") (indicatorY ++ "^")
+    | otherwise = getConfusChars xs ys charsX charsY (indicatorX ++ "-") (indicatorY ++ "-")
```
This function is a bit hard to read. I think it would be good to split it into two:

```
diffIndicator :: String -> String -> String
diffIndicator x y = zipWith (\x y -> if x == y then '-' else '^') x y
```

```
differences :: Eq a => [a] -> [a] -> [(a, a)]
differences x y = catMaybes (zipWith (\x y -> if x == y then Nothing else Just (x,y)) x y)
```

It seems this function also sometimes extends the difference indicator. You could do that with a function like this:

```
-- For example:
-- >>> overlapL "abc" "-----"
-- "abc--"
-- >>> overlapL "abc" "--"
-- "ab"
overlap :: [a] -> [a] -> [a]
overlap [] ys = ys
overlap _ [] = []
overlap (x:xs) (_:ys) = x : overlap xs ys
```

Now all pieces are in place to replace it like this:

```
getConfusCharsHelp :: String -> String -> (String, String, Stirng, String)
getConfusCharsHelp x y = (x', y', overlap di ('-' <$ x), overlap di ('-' <$ y)) where
  (x', y') = differences x y
  di = diffIndicator x y
```

Of course it is probably more readable to use `differences` and `diffIndicator` directly instead of using it via `getConfusCharsHelp` like this.

mkhakimova42 committed:
Remove unnecessary comments

mkhakimova42 committed:
Apply noughtmare's advice

- Improved readability of `GetConfusCharsHelp` and the functions it calls
- `GetConfusCharsHelp` has also been renamed to `GetConfusChars`
